### PR TITLE
*: force forbiding to use `mock.NewContext` in production code

### LINKED ISSUE #53388: remove `mock.Context` usage in production code
## Enhancement

It's better to remove the `mock.Context` usages in TiDB production code. The reason is that `mock.Context` is not designed for production code. It is designed for testing, that means some values in it are not reasonable in production and some values will be modified by the developer without any care.

We can replace some code that uses `mock.Context` with some "static" context we introduced before such as `StaticExprContext` or `DistSQLContext`

## Tasks

- #53378
- #53389
- #53441
- #53467
- #53479
- #53642
- #53671
- #53721
- #55823
- #55873
- #56038
- #56174
- #56176
- #56348
- #56355
- #56654

### PR DESCRIPTION
**Pull request by lcwangchao**
### What problem does this PR solve?

Issue Number: close #53388

### What changed and how does it work?

Add checks to avoid to use `mock.NewContext` in production code.

1. Move the function `mock.NewContext` to a standalone file which has a `//go:build !codes` header to avoid to build this function in production code. So if some one uses `mock.NewContext` in some production code, `make server` will fail.
2. Introduce a function `mock.NewContextDeprecated` and add a `// Deprecated` comment for it. This method should only be used in some legacy code.
3. Remove `mock.NewContext` usage in other production codes. 

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by YangKeao — APPROVED** [review]
LGTM

**Review by CbcWestwolf — APPROVED** [review]

File: Makefile [hunk 1]
```diff
@@ -231,9 +231,9 @@ enterprise-docker: init-submodule enterprise-prepare
 enterprise-server-build: TIDB_EDITION=Enterprise
 enterprise-server-build:
 ifeq ($(TARGET), "")
-	CGO_ENABLED=1 $(GOBUILD) -tags enterprise $(RACE_FLAG) -ldflags '$(LDFLAGS) $(CHECK_FLAG) $(EXTENSION_FLAG)' -o bin/tidb-server cmd/tidb-server/main.go
+	CGO_ENABLED=1 $(GOBUILD) -tags=codes,enterprise $(RACE_FLAG) -ldflags '$(LDFLAGS) $(CHECK_FLAG) $(EXTENSION_FLAG)' -o bin/tidb-server cmd/tidb-server/main.go
```
Comment (hawkingrei) [inline Makefile:234]: It also should add this tag for bazel.

**Review by hawkingrei — APPROVED** [review]

**Review by tangenta — APPROVED** [review]

**Pull request merged**